cli/import_map.rs

```diff
@@ -0,0 +1,2232 @@
+        match ImportMap::normalize_scope_map(scope_map, base_url) {
+          Ok(scopes_map) => scopes_map,
+          Err(err) => return Err(err),
+        }
+      }
+      None => IndexMap::new(),
+    };
+
+    let import_map = ImportMap {
+      base_url: base_url.to_string(),
+      imports: normalized_imports,
+      scopes: normalized_scopes,
+    };
+
+    Ok(import_map)
+  }
+
+  fn try_url_like_specifier(specifier: &str, base: &str) -> Option<Url> {
+    // this should never fail
+    if specifier.starts_with('/')
+      || specifier.starts_with("./")
+      || specifier.starts_with("../")
+    {
+      let base_url = Url::parse(base).unwrap();
+      let url = base_url.join(specifier).unwrap();
+      return Some(url);
+    }
+
+    if let Ok(url) = Url::parse(specifier) {
+      if SUPPORTED_FETCH_SCHEMES.contains(&url.scheme())
+        || url.scheme() == BUILT_IN_MODULE_SCHEME
+      {
+        return Some(url);
+      }
+    }
+
+    None
+  }
+
+  // TODO: add proper error handling: https://github.com/WICG/import-maps/issues/100
+  fn normalize_specifier_key(
+    specifier_key: &str,
+    base_url: &str,
+  ) -> Option<String> {
+    // ignore empty keys
+    if specifier_key.is_empty() {
+      return None;
+    }
+
+    if let Some(url) =
+      ImportMap::try_url_like_specifier(specifier_key, base_url)
+    {
+      let url_string = url.to_string();
+      if url.scheme() == BUILT_IN_MODULE_SCHEME && url_string.contains('/') {
+        eprintln!("Invalid specifier key {:?}. Built-in module specifiers must not contain \"/\".", url_string);
+        return None;
+      }
+      return Some(url_string);
+    }
+
+    // "bare" specifier
+    Some(specifier_key.to_string())
+  }
+
+  /// Addreses returned from this method are proper URLs
+  fn normalize_addresses(
+    specifier_key: &str,
+    base_url: &str,
+    address_value: Value,
+  ) -> Vec<String> {
+    let potential_addresses: Vec<Value> = match address_value {
+      Value::String(_) => vec![address_value],
+      Value::Array(address_array) => address_array,
+      Value::Null => vec![],
+      _ => vec![],
+    };
+
+    let mut normalized_addresses: Vec<String> = vec![];
+
+    for address in potential_addresses.iter() {
+      let potential_address = match address {
+        Value::String(address) => address,
+        _ => continue,
+      };
+
+      let url =
+        match ImportMap::try_url_like_specifier(potential_address, base_url) {
+          Some(url) => url,
+          None => continue,
+        };
+
+      let url_string = url.to_string();
+      if specifier_key.ends_with('/') && !url_string.ends_with('/') {
+        eprintln!(
+          "Invalid target address {:?} for package specifier {:?}.\
+           Package address targets must end with \"/\".",
+          url_string, specifier_key
+        );
+        continue;
+      }
+
+      if url.scheme() == BUILT_IN_MODULE_SCHEME && url_string.contains('/') {
```

Comment:
Please reconsider leaving it as is - built in modules can't be resolved now anyway (it will error), but if user provides remapping of built in modules in import map they will be resolved properly (again this is 1-to-1 ported from reference implementation).